### PR TITLE
[WIP] Userservice

### PR DESCRIPTION
**Pull request by jasonrichardsmith**
- Read the contribution guidelines
- Include a reference to a related issue in this repository
https://github.com/microservices-demo/microservices-demo/issues/229
https://github.com/microservices-demo/microservices-demo/issues/223
https://github.com/microservices-demo/microservices-demo/issues/107
https://github.com/microservices-demo/microservices-demo/issues/96

- A description of the changes proposed in the pull request
This is a complete rewrite of login and accounts, so bears little resemblance to either.

**This is WIP**

### CODE REVIEW

File: api/middlewares.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+		}
+	}
+}
+
+type loggingMiddleware struct {
+	next   Service
+	logger log.Logger
+}
+
+// TODO Remove passwords from Logging.
```
Comment (idcrosby) [inline api/middlewares.go]: ? I don't see any passwords being logged

File: api/service.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
+package api
+
+// service.go contains the definition and implementation (business logic) of the
+// login service. Everything here is agnostic to the transport (HTTP).
```
Comment (idcrosby) [inline api/service.go]: users service

File: api/service.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/microservices-demo/user/db"
+	"github.com/microservices-demo/user/users"
+)
+
+var (
+	passwordSalt    = "passwordsalt"
```
Comment (idcrosby) [inline api/service.go]: Should the password salt not be unique per user?

**Comment by alex-glv** [comment]
```
expected:
body:

statusCode: 200
bodySchema:
{
   "title":"List response",
   "type":"object",
   "properties":{
      "_embedded":{
         "type":"object",
         "properties":{
            "address":{
               "type":"array",
               "items":{
                  "type":"object",
                  "properties":{
                     "number":{
                        "type":"string"
                     },
                     "street":{
                        "type":"string"
                     },
                     "city":{
                        "type":"string"
                     },
                     "postcode":{
                        "type":"string"
                     },
                     "country":{
                        "type":"string"
                     },
                     "_links":{
                        "type":"object",
                        "properties":{
                           "self":{
                              "type":"object",
                              "properties":{
                                 "href":{
                                    "type":"string"
                                 }
                              },
                              "required":[
                                 "href"
                              ]
                           },
                           "address":{
                              "type":"object",
                              "properties":{
                                 "href":{
                                    "type":"string"
                                 }
                              },
                              "required":[
                                 "href"
                              ]
                           }
                        },
                        "required":[
                           "self",
                           "address"
                        ]
                     }
                  },
                  "required":[
                     "number",
                     "street",
                     "city",
                     "postcode",
                     "country",
                     "_links"
                  ]
               }
            }
         }
      },
      "_links":{
         "type":"object"
      },
      "page":{
         "type":"object"
      }
   },
   "required":[
      "_embedded",
      "_links",
      "page"
   ]
}

actual:
statusCode: 200
headers:
    content-type: application/json; charset=utf-8
    date: Tue, 16 Aug 2016 13:11:19 GMT
    content-length: 1006
    connection: close

body:
{
   "_embedded":{
      "addresses":[
         {
            "street":"Whitelees Road",
            "number":"246",
            "country":"United Kingdom",
            "city":"Glasgow",
            "postcode":"G67 3DL",
            "_links":{
               "address":{
                  "href":"http://user/addresses/57a98d98e4b00679b4a830ad"
               },
               "self":{
                  "href":"http://user/addresses/57a98d98e4b00679b4a830ad"
               }
            }
         },
         {
            "street":"Whitelees Road",
            "number":"246",
            "country":"United Kingdom",
            "city":"Glasgow",
            "postcode":"G67 3DL",
            "_links":{
               "address":{
                  "href":"http://user/addresses/57a98d98e4b00679b4a830b0"
               },
               "self":{
                  "href":"http://user/addresses/57a98d98e4b00679b4a830b0"
               }
            }
         },
         {
            "street":"Maes-Y-Deri",
            "number":"4",
            "country":"United Kingdom",
            "city":"Aberdare",
            "postcode":"CF44 6TF",
            "_links":{
               "address":{
                  "href":"http://user/addresses/57a98d98e4b00679b4a830b3"
               },
               "self":{
                  "href":"http://user/addresses/57a98d98e4b00679b4a830b3"
               }
            }
         },
         {
            "street":"my road",
            "number":"3",
            "country":"UK",
            "city":"London",
            "postcode":"",
            "_links":{
               "address":{
                  "href":"http://user/addresses/57a98ddce4b00679b4a830d1"
               },
               "self":{
                  "href":"http://user/addresses/57a98ddce4b00679b4a830d1"
               }
            }
         }
      ]
   }
}

```

That HAL thing should be in the spec :/

File: api/middlewares.go (hunk 1 above)
Comment (jasonrichardsmith) [inline api/middlewares.go]: There was a TODO saying that should be removed, so I removed it.

File: api/service.go (hunk 3 above)
Comment (jasonrichardsmith) [inline api/service.go]: I can move this to the DB,  will update.

File: api/service.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+// NewFixedService returns a simple implementation of the Service interface,
+func NewFixedService() Service {
+	return &fixedService{}
+}
+
+type fixedService struct{}
+
+func (s *fixedService) Login(username, password string) (users.User, error) {
```
Comment (alex-glv) [inline api/service.go]: The return value doesn't match the spec.
https://github.com/microservices-demo/microservices-demo/blob/master/openapi/specs/users/user.json#L269
Should be array (with one customer?)
Seems like the spec needs fixing, though.

File: api/service.go (hunk 2 above)
Comment (jasonrichardsmith) [inline api/service.go]: Are you saying rename the package?
@idcrosby

File: api/middlewares.go (hunk 1 above)
Comment (idcrosby) [inline api/middlewares.go]: cool we can remove the TODO then

File: api/service.go (hunk 2 above)
Comment (idcrosby) [inline api/service.go]: nah, just s/login/user

File: api/service.go (hunk 4 above)
Comment (jasonrichardsmith) [inline api/service.go]: Yes spec probably needs fixing

File: api/service.go (hunk 3 above)
Comment (jasonrichardsmith) [inline api/service.go]: @idcrosby updated

**Pull request merged**